[BB-6046] Prepare `opencraft-release/maple.3` branch for edx-platform

pkulkark opened the pull request:
## Description

This PR prepares the shared `opencraft-release/maple.3` branch for our fork of https://github.com/openedx/edx-platform.

[Code-drift commits](https://github.com/openedx/edx-platform/compare/open-release/lilac.2...open-craft:opencraft-release/lilac.2)

See [spreadsheet](https://docs.google.com/spreadsheets/d/1p-u2GEy9BJrxHV_F65qxdkxBnfnXwYGIUgvNn7Wy4x4/edit#gid=0) for details on the commits

Agrendalath reviewed (commented):
@pkulkark, could you please move [this spreadsheet](https://docs.google.com/spreadsheets/d/1p-u2GEy9BJrxHV_F65qxdkxBnfnXwYGIUgvNn7Wy4x4/edit#gid=0) to the PR's description? Also, please include upstream PRs there (like we did in open-craft/configuration#193) - gathering these in one place is going to make our lives much easier.

I'm copy-pasting my remark from the ticket (context in the first point below):

> Isn't testing these features a part of this task, provided that it is possible to test them? Even clean cherry-picks can still be broken if something else changes (imports, underlying functions/methods, etc.).

We are missing two things:
1. The sandbox - we should be able to reuse [this one](https://manage.opencraft.com/instance/34580/) by simply changing the commit hash of the `edx-platform`. We need to fix the code first - the test pipeline indicates that some imports are indeed broken. It will prevent migrations from running.
1. A way to test it. We should perform the [instance checklist](https://handbook.opencraft.com/en/latest/processes/instance_checklist/), but we should also verify that these cherry-picked features work correctly in Maple. If a client-specific change will be too complex, then we can let client owners know that they will need to test in a specific Stage environment. Does it sound reasonable?

cc: @navinkarkera

pkulkark commented:
@Agrendalath I've added the commits in the description. And I've fixed the imports. I have updated the commit hash and started a new appserver deployment on the [test sandbox](https://manage.opencraft.com/instance/34580/edx-appserver/23551/). I will post an update once it's ready for testing.

pkulkark commented:
@Agrendalath The [test sandbox](https://manage.opencraft.com/instance/34580/edx-appserver/23551/) was deployed successfully with the edx-platform commit set to this PR branch. I also completed the [instance checklist](https://handbook.opencraft.com/en/latest/processes/instance_checklist/) and everything seems to be working fine. I just realised however, that the configuration branch for the sandbox was set to "viadanna/maple.3" and not the one from https://github.com/open-craft/configuration/pull/193. I'm not sure if we need to test with the configuration repo branch as well or would that fall under the scope of BB-6112? CC @navinkarkera

Agrendalath reviewed (changes requested):
@pkulkark, I:
1. [Updated the description](https://www.diffchecker.com/ZnC0wFug).
2. Dropped incorrect commits from the last 5 commits in the table and cherry-picked correct references.
3. Rebased this branch to resolve conflicts.

I'm leaving my initial review below.

## General notes:
1. Please use the `git commit -x` every time. It helps us to determine where the commit came from because otherwise there is no reference to the original commit in the commit body. It might not sound too useful, but GitHub shows all repositories in which a specific commit, so it's easy to verify which named releases included it. This applies to multiple commits from this PR.
1. Please try to avoid merging the upstream branch into the PR (i.e. c962f0f3e3b6f5e3cc4714b08c2a7707aadec3fa). This is tricky to handle when we're cherry-picking things. Let's use rebase instead - I rebased this branch on `open-release/maple.master` to resolve conflicts and to drop this commit.
1. Please always cherry-pick features from either the backport PR or an upstream one. Some of these cherry-picked changes contain different diffs than we are proposing/have merged to the upstream. I haven't changed how many of these could affect the functionality, but we shouldn't add code drift to our code drift. I've added a table below with my remarks.

## Commit-specific notes:
| Ticket | Commit | Expected commit | Notes |
| -------- | ---------- | ----------------------- | -------- |
| [BB-4951](https://tasks.opencraft.com/browse/BB-4951) | openedx/edx-platform@1a218245371e7a9d3e73ceab8e49bfd7f3922fe7 | openedx/edx-platform@373f3285ea5e43da92ef57e8b940eec4bc1dbe0c | Why would we want to cherry-pick a merge commit? |
| [BB-5815](https://tasks.opencraft.com/browse/BB-5815) | 6ca5a4539b10f117af602a25f844a5c73d3856e5 | https://github.com/edx/edx-platform/pull/27857 | These commit messages do not contain anything special, so let's just squash them on the original branch, and cherry-pick a single commit from there. Also, this branch is conflicting with `master`, so we should resolve it there to avoid adding the redundant f371e32ee305f5c997617625fdd58f15dad92479. |
| [FAL-2248](https://tasks.opencraft.com/browse/FAL-2248) | bd0e5764bfba281e963d2504af0b9cda0ced71eb | a26bb857fd7b03c797534d93fe2bf26b9b8698d2 | Why would we want to cherry-pick a merge commit? |
| [BB-5271](https://tasks.opencraft.com/browse/BB-5271) | 406b0ab0c12397754c594af4fd6c361cf48ec5e8 | dd69d8e54d5b43b869835d0f877744e8776d4849 | Why would we want to cherry-pick a merge commit? |
| [BB-5819](https://tasks.opencraft.com/browse/BB-5819) | 7da10ab3ec2c6482c875a03507e952cf8d0fa370 | https://github.com/openedx/edx-platform/pull/29850 | Let's squash first, and then cherry-pick. |
| [BB-5429](https://tasks.opencraft.com/browse/BB-5429) | a20514c351c5a9aeed665e33141fbaba2edea371 | https://github.com/openedx/edx-platform/pull/30314 | Let's squash first, and then cherry-pick. |

## Ticket-specific notes:
1. BB-5849: I believe you've accidentally picked the name from the fixup commit in the cherry-pick. We should use the commit message from the base one, as it preserves the context. The same remark applies to https://github.com/openedx/edx-platform/pull/30373. Also, we don't have any ticket to track this backport.
1. BB-4381: are we planning to backport this to Maple? I don't see any open tickets, so it'll likely be forgotten as soon as we merge this PR.
1. FAL-867: I can see that [we have closed it](https://github.com/openedx/edx-platform/pull/23735#issuecomment-853484606), as we are no longer working with this client. CC: @pomegranited, as you might have [more context](https://tasks.opencraft.com/browse/FAL-867?focusedCommentId=248366&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-248366) on this item.
1. BB-5815: this entry is duplicated and uses two different commits. Please remove one of them after addressing my "Commit-specific notes".
1. BB-5819: this ticket is closed. How are we tracking the upstream PR?
1. SE-5299: this ticket is closed. How are we tracking upstream PRs?
1. BB-6041: should we create a follow-up for the backport?
1. BB-1389: once we resolve BB-5815, we can squash the leftovers from f371e32ee305f5c997617625fdd58f15dad92479 into this cherry-pick.

Agrendalath commented:
@pkulkark,
> I just realised however, that the configuration branch for the sandbox was set to "viadanna/maple.3" and not the one from https://github.com/open-craft/configuration/pull/193. I'm not sure if we need to test with the configuration repo branch as well or would that fall under the scope of BB-6112? CC @navinkarkera

Good catch. Regardless of which ticket will include testing these changes, we need to test them anyway before merging this. I changed the commit to the HEAD of the branch\* from https://github.com/open-craft/configuration/pull/193, updated its settings to a new template (we are preparing Ocim Maple support in open-craft/opencraft/pull#901), and spawned a new AppServer,

\* I've also found an odd bug (feature?) in Ocim: [SE-5566](https://tasks.opencraft.com/browse/SE-5566) (cc: @0x29a).

pomegranited commented:
> 3. FAL-867: I can see that https://github.com/openedx/edx-platform/pull/23735#issuecomment-853484606, as we are no longer working with this client. CC: @pomegranited, as you might have [more context](https://tasks.opencraft.com/browse/FAL-867?focusedCommentId=248366&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-248366) on this item.

I've [commented on FAL-867](https://tasks.opencraft.com/browse/FAL-867?focusedCommentId=248981&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-248981), but TL;DR we could make a case to upstream this now given the CC/community changes. However without a client who needs it, it's questionable whether we should. So I guess, remove it, and if anyone notices, we resurrect and upstream it?

pkulkark commented:
@Agrendalath Thanks so much for the additional help and cleanup. I initially started cherrypicking the commits from our internal lilac.2 branch, that's why it had merge commits and some that didn't match the upstream changes. I later tracked down the upstream commits and replaced it but I might've missed some.

Anyway here is what I've done now:

- I squashed the commits for https://github.com/openedx/edx-platform/pull/27857 ([BB-5815](https://tasks.opencraft.com/browse/BB-5815)) and https://github.com/openedx/edx-platform/pull/29850 ([BB-5819](https://tasks.opencraft.com/browse/BB-5815)) and cherry-picked them.
- [I pinged the owner of BB-5429](https://tasks.opencraft.com/browse/BB-5429?focusedCommentId=249009&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-249009) to squash the commits for https://github.com/openedx/edx-platform/pull/30314. Will cherry-pick once that's done.
- I created [BB-6262](https://tasks.opencraft.com/browse/BB-6262) to track the upstream PRs related to Esme (i.e. those related to [BB-5849](https://tasks.opencraft.com/browse/BB-5849), [BB-5819](https://tasks.opencraft.com/browse/BB-5819) and [BB-6041](https://tasks.opencraft.com/browse/BB-6041))
- I've asked the owners of BB-4381 and SE-5299 to create follow-up tasks to track the upstream PRs.

There were cases where either some code changes were missed to be upstreamed or the upstream PR were merged to master (but not backported to maple), which led to the tickets being marked as closed. There are still some upstream PRs which are yet to be reviewed and so can't be backported to maple yet. We'll need a better process for this.

Also the [test sandbox](https://manage.opencraft.com/instance/34580/edx-appserver/23586/) was deployed successfully. I'll go through the instance checklist once again on it. But for a more detailed testing (like checking if the changes in the commits work or not) I think a follow-up task would be better, since those will involve the changes in the configuration repo as well. What do you think?

pkulkark commented:
@Agrendalath I went through the instance checklist on the latest test sandbox, but it looks like Account MFE is not working. It was working on the previous one (i.e. before changing the configuration version) so it looks like this is something that should be addressed in [BB-6112](https://tasks.opencraft.com/browse/BB-6112)?

0x29a commented:
>I went through the instance checklist on the latest test sandbox, but it looks like Account MFE is not working. It was working on the previous one (i.e. before changing the configuration version) so it looks like this is something that should be addressed in [BB-6112](https://tasks.opencraft.com/browse/BB-6112)?

@pkulkark, I checked what's wrong, and it looks like `MFES` variable from the instance's `configuration_extra_settings` is missing some elements.

Current:
```yml
MFES:
  - name: gradebook
    repo: frontend-app-gradebook
    public_path: /gradebook/
```
Should be:
```yml
MFES:
  - name: profile
    repo: frontend-app-profile
    public_path: "/profile/"
  - name: gradebook
    repo: frontend-app-gradebook
    public_path: "/gradebook/"
  - name: account
    repo: frontend-app-account
    public_path: "/account/"
  - name: learning
    repo: frontend-app-learning
    public_path: "/learning/"
```

Agrendalath commented:
@pkulkark, @0x29a, this is intentional. We are not deploying other MFEs than the gradebook, because they are student-facing and we cannot theme them yet. I believe I've mentioned this in the discovery ticket, so it should be in the discovery doc.

pkulkark commented:
@Agrendalath
> this is intentional. We are not deploying other MFEs than the gradebook, because they are student-facing and we cannot theme them yet. I believe I've mentioned this in the discovery ticket, so it should be in the discovery doc.

I thought that was only for learning MFE. Anyway this is breaking the accounts page (results in 404). Is there some configuration that forces to not use the MFE for the accounts endpoint? CC @navinkarkera @0x29a
